balancer/catalyst/catalyst_balancer.go

```diff
@@ -0,0 +1,272 @@
+
+	// 1. Has Stream and Is Local and Isn't Overloaded
+	localHasStreamNotOverloaded := []ScoredNode{}
+	for _, node := range scoredNodes {
+		if node.GeoScore == 2 && node.HasStream(streamID) && node.GetLoadScore() == 2 {
+			localHasStreamNotOverloaded = append(localHasStreamNotOverloaded, node)
+		}
+	}
+	if len(localHasStreamNotOverloaded) > 0 { // TODO: Should this be > 1 or > 2 so that we can ensure there's always some randomness?
+		return localHasStreamNotOverloaded
+	}
+
+	// 2. Is Local and Isn't Overloaded
+	localNotOverloaded := []ScoredNode{}
+	for _, node := range scoredNodes {
+		if node.GeoScore == 2 && node.GetLoadScore() == 2 {
+			localNotOverloaded = append(localNotOverloaded, node)
+		}
+	}
+	if len(localNotOverloaded) > 0 { // TODO: Should this be > 1 or > 2 so that we can ensure there's always some randomness?
+		return localNotOverloaded
+	}
+
+	// 3. Weighted least-bad option
+	for i, node := range scoredNodes {
+		node.Score += node.GeoScore
+		node.Score += int64(node.GetLoadScore())
+		if node.HasStream(streamID) {
+			node.Score += 2
+		}
+		scoredNodes[i] = node
+	}
+
+	sort.Slice(scoredNodes, func(i, j int) bool {
+		return scoredNodes[i].Score > scoredNodes[j].Score
+	})
+	if len(scoredNodes) < numNodes {
+		return scoredNodes
+	}
+
+	return scoredNodes[:numNodes]
+}
+
+func (c *CataBalancer) MistUtilLoadSource(ctx context.Context, stream, lat, lon string) (string, error) {
+	for _, node := range c.Nodes {
+		removeOldStreams(c.IngestStreams[node.Name])
+		if _, ok := c.IngestStreams[node.Name]; ok {
+			if _, ok := c.IngestStreams[node.Name][stream]; ok {
+				return node.DTSC, nil
+			}
+		}
+	}
+	return "", fmt.Errorf("no node found for ingest stream: %s", stream)
+}
+
+func (c *CataBalancer) UpdateNodes(id string, nodeMetrics NodeMetrics) {
+	c.metricsLock.Lock()
+	defer c.metricsLock.Unlock()
+	log.LogNoRequestID("catabalancer updatenodes", "id", id, "ram", nodeMetrics.RAMUsagePercentage, "cpu", nodeMetrics.CPUUsagePercentage)
+	if _, ok := c.Nodes[id]; !ok {
+		log.LogNoRequestID("catabalancer updatenodes node not found", "id", id)
+		return
+	}
+	c.NodeMetrics[id] = nodeMetrics
+}
+
+var streamTimeout = 5 * time.Second // should match how often we send the update messages
+
+func (c *CataBalancer) UpdateStreams(id string, streamID string, isIngest bool) {
+	c.streamsLock.Lock()
+	defer c.streamsLock.Unlock()
+	if _, ok := c.Nodes[id]; !ok {
+		log.LogNoRequestID("catabalancer updatestreams node not found", "id", id)
+		return
+	}
+	// remove old streams
+	removeOldStreams(c.Streams[id])
+	removeOldStreams(c.IngestStreams[id])
+
+	if isIngest {
+		if c.IngestStreams[id] == nil {
+			c.IngestStreams[id] = make(Streams)
+		}
+		c.IngestStreams[id][streamID] = Stream{ID: streamID, LastSeen: time.Now()}
+		return
+	}
+	if c.Streams[id] == nil {
+		c.Streams[id] = make(Streams)
+	}
+	c.Streams[id][streamID] = Stream{ID: streamID, LastSeen: time.Now()}
+}
+
+func removeOldStreams(streams Streams) {
```

Comment:
Ah thanks good spot, will fix that!

Reply:
I really wanted to avoid having to acquire locks for the codepaths associated with end user requests, so I've removed the calls to `removeOldStreams` in those cases and instead just check if the stream is expired based on `LastSeen`. Does that sound ok?
We'll be getting calls to `UpdateStreams` every 5s for each other node where we do call `removeOldStreams` so we'll be cleaning up old memory that way.

Reply:
Ok actually that's no good, go's concurrent read and write detector will kick in. Already seen it happen in staging. So i'll have to just minimise the amount of time we need to hold the locks.